tests/blackbox.py

```diff
@@ -130,72 +196,176 @@ def blackbox_pyteal(subr: BlackboxWrapper, mode: Mode) -> Callable[..., Expr]:
-    ), "please provide input_types in your @Subroutine annotation (crucial for generating proper end-to-end testable PyTeal)"
-
-    subdef = subr.subroutine.subroutine
-    arg_names = subdef.arguments()
-
-    def arg_prep_n_call(i, p):
-        name = arg_names[i]
-        by_ref = name in subdef.by_ref_args
-        arg_expr = Txn.application_args[i] if mode == Mode.Application else Arg(i)
-        if p == TealType.uint64:
-            arg_expr = Btoi(arg_expr)
-        prep = None
-        arg_var = arg_expr
-        if by_ref:
-            arg_var = ScratchVar(p)
-            prep = arg_var.store(arg_expr)
-        return prep, arg_var
-
-    def subr_caller():
-        preps_n_calls = [*(arg_prep_n_call(i, p) for i, p in enumerate(input_types))]
-        preps, calls = zip(*preps_n_calls) if preps_n_calls else ([], [])
-        preps = [p for p in preps if p]
-        invocation = subr(*calls)
-        if preps:
-            return Seq(*(preps + [invocation]))
-        return invocation
-
-    def make_return(e):
-        if e.type_of() == TealType.uint64:
-            return e
-        if e.type_of() == TealType.bytes:
-            return Len(e)
-        if e.type_of() == TealType.anytype:
-            x = ScratchVar(TealType.anytype)
-            return Seq(x.store(e), Int(1337))
-        # TealType.none:
-        return Seq(e, Int(1337))
-
-    def make_log(e):
-        if e.type_of() == TealType.uint64:
-            return Log(Itob(e))
-        if e.type_of() == TealType.bytes:
-            return Log(e)
-        return Log(Bytes("nada"))
-
-    if mode == Mode.Signature:
-
-        def approval():
-            return make_return(subr_caller())
-
-    else:
+    return BlackboxPyTealer(subr, mode).program()
+
+
+class BlackboxPyTealer:
+    def __init__(self, subr: BlackboxWrapper, mode: Mode):
+        input_types = subr.input_types
+        assert (
+            input_types is not None
+        ), "please provide input_types in your @Subroutine annotation (crucial for generating proper end-to-end testable PyTeal)"
+
+        self.subr, self.mode, self.input_types = subr, mode, input_types
+        match subr.subroutine:
+            case SubroutineFnWrapper():
+                approval = self._handle_SubroutineFnWrapper()
+            case ABIReturnSubroutine():
+                approval = self._handle_ABIReturnSubroutine()
+            case _:
+                raise AssertionError(
+                    f"Cannot produce Blackbox pyteal for provided subroutine of type {type(subr.subroutine)}"
+                )
+
+        setattr(approval, "__name__", f"sem_{mode}_{subr.name()}")
+        self._pyteal_lambda: Callable[..., Expr] = approval
+
+    def is_abi(self) -> bool:
+        return isinstance(self.subr.subroutine, ABIReturnSubroutine)
+
+    def abi_argument_types(self) -> list[None | algosdk.abi.ABIType]:
+        if not self.is_abi():
+            raise NotImplementedError("this {type(self)} is not A.B.I. compatible")
```

Comment:
@tzaffi Minor:  I _think_ the PR intends for an f-string here.  If that's the case, there's another occurrence in the same file.

```suggestion
            raise NotImplementedError(f"this {type(self)} is not A.B.I. compatible")
```